reposerver/repository/repository.go

```diff
@@ -1294,27 +1294,39 @@ func getPluginEnvs(envVars *v1alpha1.Env, q *apiclient.ManifestRequest, creds gi
 
 	if q.ApplicationSource.Plugin != nil {
 		pluginEnv := q.ApplicationSource.Plugin.Env
-		for i, j := range pluginEnv {
-			pluginEnv[i].Value = parsedEnv.Envsubst(j.Value)
+		for _, entry := range pluginEnv {
+			newValue := parsedEnv.Envsubst(entry.Value)
+			if legacy {
```

Comment:
@leoluz suggested we consider removing the legacy behavior completely in 2.4. I think it's worth considering moving aggressively, since it's a security enhancement. I've added a note to Thursday's contributor meeting to discuss. https://docs.google.com/document/d/1xkoFkVviB70YBzSEa4bDnu-rUZ1sIFtwKKG1Uw8XsY8/edit#

Reply:
> But this will also break published plugins.

True. I think it's only justified if we think the security benefit is significant and we can communicate the change effectively.

> Have you considered building the env array in reverse order ? This way a user cannot override an existing variable.

I don't think that's sufficient to resolve the security concern. Tooling in the CMP might pick up env vars that aren't explicitly set elsewhere and allow malicious behavior. 

> I am not sure if this is really intended.

I think allowing the user to override CMP-side vars was intentional, but I think it was a bad idea.

By prefixing the env var names, I think we'll make it so that the array order doesn't matter - there should be no overlap in env var names.

Reply:
The side note was about the fact that both the repo server and the cmp server appended their environment variables to the environment used in the call (priority for the repoServer. For example, the value of PATH is the wrong one). I think it is a side effect of the fact that `getPluginEnvs`  is also used for plugin loaded in the repoServer.

Reply:
Oh wait. So the repo-server PATH var (and other vars) end up taking precedence over the sidecar? If so, yikes.

Reply:
I have made a real test adding a variable TEST to both containers with value either repo-server or cmp-server and I can confirm what I inferred from the code TEST value was repo-server in the plugin. I can add a boolean variable remote to distinguish between the behavior for https://github.com/argoproj/argo-cd/blob/08009211ff209ea20ee69155cd80f9397545c8e0/reposerver/repository/repository.go#L1255 and the behavior required for sidecars.

Reply:
Agreed, separate PR makes sense.

Reply:
I have pushed #9397